### PR TITLE
Drivetrain

### PR DESCRIPTION
**Pull request by codingvic**
Doper drivetrain code

### CODE REVIEW

File: src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/Drivetrain.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            ki = 0.0
+            kd = 0.0
+        }
+    }
+
+    val rightPosition get() = right.position
+    val leftPosition get() = left.position
+    val position get() = (rightPosition + leftPosition) / 2
+
+
```
Comment (gnawinganimal) [inline src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/Drivetrain.kt:51]: No double spaces plez

File: src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/Drivetrain.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        right.setPercentOutput(0.0)
+        left.setPercentOutput(0.0)
+    }
+
+    fun zeroEncoders() {
+        right.position = 0
+        left.position = 0
+    }
+
+}
```
Comment (gnawinganimal) [inline src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/Drivetrain.kt]: Add space at end plez

File: src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/Drivetrain.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    fun stop() {
+        right.setPercentOutput(0.0)
+        left.setPercentOutput(0.0)
+    }
+
+    fun zeroEncoders() {
+        right.position = 0
+        left.position = 0
+    }
+
```
Comment (gnawinganimal) [inline src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/Drivetrain.kt]: No space at end of class plez

File: src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/DrivetrainCommands.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                Radians / Seconds,
+                speed.convertTo(Meters / Seconds).value / WHEEL_RADIUS_METERS
+        ).convertTo(EncoderTicks(PULSES_PER_REVOLUTION.toInt()) / Seconds).value
+        val curve = MotionCurve(ticks, ticksPerSecond, 100.0, 100.0)
+
+        timer(20, 0, (curve.t7 * 1000).toLong()) {
+            drivetrain.driveToPosition(curve.d(it / 1000.0).toInt())
+        }
+    }
+}
```
Comment (gnawinganimal) [inline src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/DrivetrainCommands.kt]: Add space at end plez

File: src/main/kotlin/org/sert2521/infiniterecharge2020/utils/Timer.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        periodic(period) {
+            task()
+            elapsedTime += period
+            if (elapsedTime > runTime) {
+                cancel()
+            }
+        }
+    }.join()
+}
+
```
Comment (gnawinganimal) [inline src/main/kotlin/org/sert2521/infiniterecharge2020/utils/Timer.kt:21]: No extra space at end plez

File: src/main/kotlin/org/sert2521/infiniterecharge2020/oi/OperatorInput.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    addOption("Joystick", ControlMode.JOYSTICK)
+    addOption("Controller", ControlMode.CONTROLLER)
+}
+
+val controlMode get() = controlModeChooser.selected ?: ControlMode.JOYSTICK
+
+val primaryJoystick by lazy { Joystick(Operator.PRIMARY_STICK) }
+val secondaryJoystick by lazy { Joystick(Operator.SECONDARY_STICK) }
+val primaryController by lazy { XboxController(Operator.PRIMARY_CONTROLLER) }
+
```
Comment (gnawinganimal) [inline src/main/kotlin/org/sert2521/infiniterecharge2020/oi/OperatorInput.kt]: No extra space at end plez

File: src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/DrivetrainPreferences.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+import edu.wpi.first.wpilibj.Preferences
+
+val driveSpeedScalar get() = Preferences.getInstance().getDouble("drive_speed_scalar", 1.0)
+
+// Number of encoder pulses for one revolution of the wheel
+// TODO: Tune on carpet
+const val PULSES_PER_REVOLUTION = 7442.0
+const val WHEEL_RADIUS = 3.0
+const val WHEEL_RADIUS_METERS = 0.0254
```
Comment (gnawinganimal) [inline src/main/kotlin/org/sert2521/infiniterecharge2020/drivetrain/DrivetrainPreferences.kt]: Add space at end plez

**Comment by mckelveygreg** [comment]
Don't forget to resolve comments, and please write good commit messages! You'll thank yourself later

**Comment by will-hou** [comment]
The `controlMode` SendableChooser still isn't working and I'm not too sure why. Will compare again with the code from BunnyBots and continue to troubleshoot on Friday.

**Comment by mckelveygreg** [comment]
hmm.. what does the `Update branch` button do?
This whole protections thing is new to me...

**Comment by gnawinganimal** [comment]
I think update branch just merges from master

**Review by gnawinganimal — APPROVED** [review]
Pretty good if I do say so myself

**Pull request merged**